File: Module1/T1.1/submission/src/integration.py
import csv
from datetime import datetime
from decimal import Decimal, InvalidOperation
import json
from dataclasses import dataclass, asdict
from pathlib import Path
import sys
from uuid import uuid4
import structlog

DATE_FORMATS: tuple[str, ...] = (
    "%Y-%m-%d",
    "%m/%d/%Y",
    "%b %d, %Y"
)


def normalize_date(date: str) -> str | None:
    clean_date = date.strip()

    for format in DATE_FORMATS:
        try:
            parsed_date = datetime.strptime(clean_date, format)
            return parsed_date.date().isoformat()
        except ValueError:
            continue
        
    return None


def _is_refund(cleaned_amount: str) -> bool:
    if cleaned_amount[0] == "(" and cleaned_amount[-1] == ")":
        return True
    return False


def normalize_amount(amount: str) -> int | None:
    cleaned_amount = amount.strip()

    if cleaned_amount == "" or cleaned_amount.upper() == "N/A":
        return None

    is_refund = _is_refund(cleaned_amount=cleaned_amount)
    if is_refund:
        cleaned_amount = cleaned_amount[1:-1]
    
    cleaned_amount = (
        cleaned_amount
        .replace("$", "")
        .replace(",", "")
        .replace(" ", "")
        )
    
    try:
        decimal_amount = Decimal(cleaned_amount)
    except InvalidOperation:
        return None

    cents = decimal_amount * 100

    amount_cents = int(cents)

    if is_refund:
        return -abs(amount_cents)
    
    return amount_cents


def normalize_currency(currency: str) -> str | None:
    cleaned_currency = currency.strip()
    if cleaned_currency == "" or cleaned_currency == "$":
        return "USD"
    
    return cleaned_currency.upper()


def normalize_merchant(merchant: str) -> str | None:
    return " ".join(merchant.split())

@dataclass(frozen=True, slots=True)
class CleanRow():
    row_id: str
    date: str
    amount_cents: int | None
    currency:str | None
    merchant: str | None


@dataclass(frozen=True, slots=True)
class ErrorRow:
    row_id: str
    reason: str

@dataclass
class Output:
    clean: list[CleanRow]
    errors: list[ErrorRow]


def normalize_rows(input_path: str, output_path: str) -> Output:
    clean: list[CleanRow] = []
    errors: list[ErrorRow] = []

    with open(input_path, newline="", encoding="utf-8") as input_file:
        reader = csv.DictReader(input_file)

        for row in reader:
            row_id = row.get("row_id") or ""
            normalized_date = normalize_date(row.get("date") or "")
            normalized_amount = normalize_amount(row.get("amount") or "")

            if normalized_date is None:
                errors.append(
                    ErrorRow(
                        row_id=row_id,
                        reason="unparseable_date",
                    )
                )
                continue

            if normalized_amount is None:
                errors.append(
                    ErrorRow(
                        row_id=row_id,
                        reason="unparseable_amount",
                    )
                )
                continue

            clean.append(
                CleanRow(
                    row_id=row_id,
                    date=normalized_date,
                    merchant=normalize_merchant(
                        row.get("merchant") or ""
                    ),
                    amount_cents=normalized_amount,
                    currency=normalize_currency(
                        row.get("currency") or ""
                    ),
                )
            )

    output = Output(clean=clean, errors=errors)

    # This must be "w", not "a".
    with open(output_path, "w", encoding="utf-8") as output_file:
        json.dump(asdict(output), output_file, indent=2)
        output_file.write("\n")

    return output



def main() -> None:
    if len(sys.argv) != 3:
        raise SystemExit("inegration.py, input csv, output json")
    log_path = Path(__file__).parent.parent / "run.log"

    with open(log_path, "w", encoding="utf-8") as log_file:
        structlog.configure(
            processors=[structlog.processors.JSONRenderer(),],
            logger_factory=structlog.PrintLoggerFactory(
                file = log_file
            ),
        )

        log = structlog.get_logger(
        correlation_id = str(uuid4())   
        )

        log.info("run_started")

        output = normalize_rows(
            sys.argv[1],
            sys.argv[2],
        )

        log.info(
            "run completed",
            clean_count = len(output.clean),
            error_count = len(output.errors),
        )

if __name__ == "__main__":
    main()

File: Module1/T1.2/submission/src/integration.ts
import { parse } from "csv-parse/sync";
import { randomUUID } from "node:crypto";
import { closeSync, openSync, readFileSync, writeFileSync, writeSync } from "node:fs";
import { dirname, join } from "node:path";
import { fileURLToPath, pathToFileURL } from "node:url";

const DATE_FORMATS = ["iso", "us-slash", "month-name"] as const;

const MONTH_ABBREVIATIONS: Record<string, number> = {
  jan: 1,
  feb: 2,
  mar: 3,
  apr: 4,
  may: 5,
  jun: 6,
  jul: 7,
  aug: 8,
  sep: 9,
  oct: 10,
  nov: 11,
  dec: 12,
};

interface DateParts {
  year: number;
  month: number;
  day: number;
}

function daysInMonth(year: number, month: number): number {
  return new Date(Date.UTC(year, month, 0)).getUTCDate();
}

function isValidDate(year: number, month: number, day: number): boolean {
  if (month < 1 || month > 12) return false;
  if (day < 1) return false;
  return day <= daysInMonth(year, month);
}

function toIso(parts: DateParts): string {
  const y = String(parts.year).padStart(4, "0");
  const m = String(parts.month).padStart(2, "0");
  const d = String(parts.day).padStart(2, "0");
  return `${y}-${m}-${d}`;
}

function parseIsoDate(value: string): DateParts | null {
  const match = /^(\d{4})-(\d{2})-(\d{2})$/.exec(value);
  if (!match) return null;
  const year = Number(match[1]);
  const month = Number(match[2]);
  const day = Number(match[3]);
  if (!isValidDate(year, month, day)) return null;
  return { year, month, day };
}

function parseUsSlashDate(value: string): DateParts | null {
  const match = /^(\d{1,2})\/(\d{1,2})\/(\d{4})$/.exec(value);
  if (!match) return null;
  const month = Number(match[1]);
  const day = Number(match[2]);
  const year = Number(match[3]);
  if (!isValidDate(year, month, day)) return null;
  return { year, month, day };
}

function parseMonthNameDate(value: string): DateParts | null {
  const match = /^([A-Za-z]{3}) (\d{1,2}), (\d{4})$/.exec(value);
  if (!match) return null;
  const month = MONTH_ABBREVIATIONS[match[1]!.toLowerCase()];
  if (month === undefined) return null;
  const day = Number(match[2]);
  const year = Number(match[3]);
  if (!isValidDate(year, month, day)) return null;
  return { year, month, day };
}

function parseByFormat(format: (typeof DATE_FORMATS)[number], value: string): DateParts | null {
  switch (format) {
    case "iso":
      return parseIsoDate(value);
    case "us-slash":
      return parseUsSlashDate(value);
    case "month-name":
      return parseMonthNameDate(value);
  }
}

export function normalize_date(date: string): string | null {
  const cleanDate = date.trim();

  for (const format of DATE_FORMATS) {
    const parts = parseByFormat(format, cleanDate);
    if (parts !== null) {
      return toIso(parts);
    }
  }

  return null;
}

function _is_refund(cleanedAmount: string): boolean {
  return cleanedAmount.startsWith("(") && cleanedAmount.endsWith(")");
}

export function normalize_amount(amount: string): number | null {
  let cleanedAmount = amount.trim();

  if (cleanedAmount === "" || cleanedAmount.toUpperCase() === "N/A") {
    return null;
  }

  const isRefund = _is_refund(cleanedAmount);
  if (isRefund) {
    cleanedAmount = cleanedAmount.slice(1, -1);
  }

  cleanedAmount = cleanedAmount.replaceAll("$", "").replaceAll(",", "").replaceAll(" ", "");

  if (cleanedAmount === "") {
    return null;
  }

  const value = Number(cleanedAmount);
  if (!Number.isFinite(value)) {
    return null;
  }

  // Number(x)*100 can land a hair off an integer (binary float rounding),
  // e.g. 44999.999999999996 instead of 45000 — Math.round absorbs that.
  const amountCents = Math.round(value * 100);

  if (isRefund) {
    return -Math.abs(amountCents);
  }

  return amountCents;
}

export function normalize_currency(currency: string): string | null {
  const cleanedCurrency = currency.trim();
  if (cleanedCurrency === "" || cleanedCurrency === "$") {
    return "USD";
  }

  return cleanedCurrency.toUpperCase();
}

export function normalize_merchant(merchant: string): string | null {
  return merchant.trim().split(/\s+/).join(" ");
}

export interface CleanRow {
  row_id: string;
  date: string;
  amount_cents: number | null;
  currency: string | null;
  merchant: string | null;
}

export interface ErrorRow {
  row_id: string;
  reason: string;
}

export interface Output {
  clean: CleanRow[];
  errors: ErrorRow[];
}

export function normalize_rows(inputPath: string, outputPath: string): Output {
  const clean: CleanRow[] = [];
  const errors: ErrorRow[] = [];

  const inputContent = readFileSync(inputPath, "utf-8");
  const rows: Record<string, string>[] = parse(inputContent, {
    columns: true,
  });

  for (const row of rows) {
    const rowId = row.row_id ?? "";
    const normalizedDate = normalize_date(row.date ?? "");
    const normalizedAmount = normalize_amount(row.amount ?? "");

    if (normalizedDate === null) {
      errors.push({
        row_id: rowId,
        reason: "unparseable_date",
      });
      continue;
    }

    if (normalizedAmount === null) {
      errors.push({
        row_id: rowId,
        reason: "unparseable_amount",
      });
      continue;
    }

    clean.push({
      row_id: rowId,
      date: normalizedDate,
      amount_cents: normalizedAmount,
      currency: normalize_currency(row.currency ?? ""),
      merchant: normalize_merchant(row.merchant ?? ""),
    });
  }

  const output: Output = { clean, errors };

  // This must truncate, not append.
  writeFileSync(outputPath, `${JSON.stringify(output, null, 2)}\n`, "utf-8");

  return output;
}

export function placeholder(): boolean {
  return true;
}

function main(): void {
  if (process.argv.length !== 4) {
    console.error("integration.ts, input csv, output json");
    process.exit(1);
  }

  const [, , inputPath, outputPath] = process.argv;
  const scriptDir = dirname(fileURLToPath(import.meta.url));
  const logPath = join(scriptDir, "..", "run.log");
  const correlationId = randomUUID();

  const logFd = openSync(logPath, "w");
  const log = (event: string, fields: Record<string, unknown> = {}): void => {
    const line = JSON.stringify({ correlation_id: correlationId, ...fields, event });
    writeSync(logFd, `${line}\n`);
  };

  log("run_started");

  const output = normalize_rows(inputPath as string, outputPath as string);

  log("run completed", {
    clean_count: output.clean.length,
    error_count: output.errors.length,
  });

  closeSync(logFd);
}

if (process.argv[1] && import.meta.url === pathToFileURL(process.argv[1]).href) {
  main();
}
